src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataBlockService.cs

```diff
@@ -122,9 +122,32 @@ public class DataBlockService : IDataBlockService
                 {
                     var viewModel = _mapper.Map<DataBlockViewModel>(dataBlock);
 
+                    var subjectId = dataBlock.Query.SubjectId;
+
+                    var releaseContentBlock = await _context.ReleaseContentBlocks
+                        .Include(rcb => rcb.Release)
+                        .SingleOrDefaultAsync(rcb => rcb.ContentBlockId == id);
+
+                    var releaseId = releaseContentBlock.ReleaseId;
+
+                    viewModel.DataSetId = await _context.ReleaseFiles.Include(rf => rf.File)
+                        .Where(rf =>
+                            rf.ReleaseId == releaseId &&
+                            rf.File.SubjectId == subjectId &&
+                            rf.File.Type == FileType.Data)
+                        .Select(rf => rf.File.Id).SingleAsync();
+
+                    viewModel.DataSetName = await _context.ReleaseFiles.Include(rf => rf.File)
+                        .Where(rf =>
+                            rf.ReleaseId == releaseId &&
+                            rf.File.SubjectId == subjectId &&
+                            rf.File.Type == FileType.Data)
+                        .Select(rf => rf.Name).SingleAsync();
+
                     var featuredTable = await _context.FeaturedTables.SingleOrDefaultAsync(
                         ft => ft.DataBlockId == dataBlock.Id);
```

Comment:
You could abstract out the common predicate for getting the `ReleaseFile` as a `Queryable` here which isn't executed yet. You'd end up with this which avoids duplicating the predicate twice:

```
var releaseFile = _context.ReleaseFiles
    .Where(rf =>
        rf.ReleaseId == releaseId &&
        rf.File.SubjectId == subjectId &&
        rf.File.Type == FileType.Data);

viewModel.DataSetId = await releaseFile
    .Select(rf => rf.FileId)
    .SingleAsync();

viewModel.DataSetName = await releaseFile
    .Select(rf => rf.Name)
    .SingleAsync() ?? "";
```

The problem with all this is that including fetching the release id above and getting the data set id and name here, you have 3 queries in total. This optimisation doesn't change that.

If your comfortable with the syntax, you can do everything in one query. This avoids a separate call to get the release id, eliminating what you've got above.

```
var releaseFile = await _context.ReleaseContentBlocks
    .Join(_context.ReleaseFiles,
        rcb => rcb.ReleaseId,
        rf => rf.ReleaseId,
        (rcb, rf) => new { rcb, rf })
    .Where(tuple =>
        tuple.rcb.ContentBlockId == id &&
        tuple.rf.File.SubjectId == subjectId &&
        tuple.rf.File.Type == FileType.Data)
    .Select(t => new
    {
        DataSetId = t.rf.FileId,
        DataSetName = t.rf.Name
    })
    .SingleAsync();

viewModel.DataSetId = releaseFile.DataSetId;
viewModel.DataSetName = releaseFile.DataSetName ?? "";
```

Reply:
As we can get the `releaseId` from the result of `GetReleaseContentBlock` - see my other comment - I don't believe this join is necessary.

I think instead we can just do this?

```
                    viewModel.DataSetName = await _context.ReleaseFiles
                        .Where(rf =>
                            rf.ReleaseId == releaseId &&
                            rf.File.SubjectId == subjectId &&
                            rf.File.Type == FileType.Data)
                        .Select(rf => rf.Name)
                        .SingleAsync();
```

Reply:
As we can get the releaseId from the result of GetReleaseContentBlock - see my other comment - I don't believe this join is necessary.

I think instead we can just do this?

```
                    viewModel.DataSetName = await _context.ReleaseFiles
                        .Where(rf =>
                            rf.ReleaseId == releaseId &&
                            rf.File.SubjectId == subjectId &&
                            rf.File.Type == FileType.Data)
                        .Select(rf => rf.Name)
                        .SingleAsync();
```

Reply:
Sorted now. As discussed with Ben, it might be worth optimizing the `GetReleaseContentBlock` method to not return the entire release